p-867 Refactoring extra params

0xverin opened the pull request:
### Context

This PR refactoring the structure of the contract, using params as parameters to match different token library.

~~We don't need to deploy each token now, just deploy the [ERC20Mapping](https://github.com/litentry/litentry-parachain/pull/2830/files#diff-d55cbd2a8d37a86353163ab51eaf135c93b65a4713046e56952c26337bfb311e) and [BRC20Mapping](https://github.com/litentry/litentry-parachain/pull/2830/files#diff-d9754d115d0f67f4b3e97b3c2eeba4ceb0485bdb90ee80a6fcdf0f657a9eb106) contracts,according to the decoded params,we can find the corresponding mapping.~~

Now just need to deploy the [TokenMapping](https://github.com/litentry/litentry-parachain/pull/2830/files#diff-ffbd58367b287b2fd9014f61f6ccffbfd6e84b451e80279d1d869ce1f5ebe985) contract.

`extra params` should be abi.encode of the token lowercase name, for example:

abi.encode('bnb'):`00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626e620000000000000000000000000000000000000000000000000000000000`

### Labels

Please apply following PR-related labels when appropriate:
- `C0-breaking`: if your change could break the existing client, e.g. API change, critical logic change 
- `C1-noteworthy`: if your change is non-breaking, but is still worth noticing for the client, e.g. reference code improvement

### How (Optional)

<!-- How were these changes implemented? -->

### Testing Evidences

```
 ./bin/litentry-cli trusted -d link-identity did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e did:litentry:evm:0x5ad727D30bbcB5A7bCdFa213B940cCFA61835121 ethereum 
```

```
./bin/litentry-cli trusted -d request-vc did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e -a "dynamic 0000000000000000000000000000000000000007 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626e620000000000000000000000000000000000000000000000000000000000"
```

```
{
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
    ],
    "id": "0xd250829993d57dd45d51c6828fde7af1b15c833621672a7cb7a2e63530aa764b",
    "type": [
        "VerifiableCredential"
    ],
    "credentialSubject": {
        "id": "did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e",
        "description": "The amount of a particular token you are holding",
        "type": "Token Holding Amount",
        "assertionText": "Dynamic(0x0000000000000000000000000000000000000007, BoundedVec([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 98, 110, 98, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 1024))",
        "assertions": [
            {
                "and": [
                    {
                        "src": "$token",
                        "op": "==",
                        "dst": "Bnb"
                    },
                    {
                        "src": "$holding_amount",
                        "op": ">=",
                        "dst": "500"
                    },
                    {
                        "src": "$holding_amount",
                        "op": "<",
                        "dst": "800"
                    }
                ]
            }
        ],
        "values": [
            true
        ],
        "endpoint": "http://localhost:9933/"
    },
    "issuer": {
        "id": "did:litentry:substrate:0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15",
        "name": "Litentry TEE Worker",
        "mrenclave": "9vnN35nXEq76rUS2ywFAgAjjh2HHCF3LjHck9SSuQEwH",
        "runtimeVersion": {
            "parachain": 9181,
            "sidechain": 107
        }
    },
    "issuanceDate": "2024-06-20T05:04:47.349866428+00:00",
    "parachainBlockNumber": 5222,
    "sidechainBlockNumber": 10432,
    "proof": {
        "created": "2024-06-20T05:04:47.351085789+00:00",
        "type": "Ed25519Signature2020",
        "proofPurpose": "assertionMethod",
        "proofValue": "1d851ab95493798f0080157776c24051ebdacdaf1734e1ec386a3614dab2ff3c19744bebb575a98aaa538acaf89087af4434cd7c4ae05fe10065f9119410820b",
        "verificationMethod": "0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15"
    },
    "credentialSchema": {
        "id": "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json",
        "type": "JsonSchemaValidator2018"
    }
}
```


<img width="1788" alt="image" src="https://github.com/litentry/litentry-parachain/assets/104152026/88339f6a-f6eb-45c8-8ad6-ebc0ec5ecf4a">

0xverin committed:
refactor TokenHoldingAmount

0xverin committed:
refactor erc20&brc20

0xverin committed:
add isStringsEqual

0xverin committed:
refactor erc20 token

0xverin committed:
refactor brc20 tokens

0xverin committed:
remove console

0xverin committed:
Merge branch 'dev' into refactor-extra-params

0xverin committed:
reset url

BillyWooo commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol:
```diff
@@ -21,62 +21,86 @@ pragma solidity ^0.8.8;
-import {TokenHoldingAmount} from "./TokenHoldingAmount.sol";
+import { TokenHoldingAmount } from "./TokenHoldingAmount.sol";
+import { Btcs } from "./brc20/Btcs.sol";
+import { Cats } from "./brc20/Cats.sol";
```
only these two? I see there are quite some others, like long, mmss, ordi, rats, sats?

BillyWooo commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol:
```diff
@@ -21,62 +21,86 @@ pragma solidity ^0.8.8;
+		} else if (Utils.isStringsEqual(decodedParams, "Cats")) {
+			tokenName = Cats.getTokenName();
+			ranges = Cats.getTokenRanges();
+		} else {
```
same problem as above. Just forget?

BillyWooo commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol:
```diff
@@ -17,34 +17,40 @@
-abstract contract ERC20 is TokenHoldingAmount {
-	mapping(uint32 => string) internal networkTokenAddresses;
+import { Ada } from "./erc20/Ada.sol";
+import { Amp } from "./erc20/Amp.sol";
```
similar problem as above. 
Do I misunderstand anything?

0xverin replied:
I added all and made some changes to the positions.

BillyWooo commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol:
```diff
@@ -96,4 +102,34 @@ abstract contract ERC20 is TokenHoldingAmount {
+			ranges = Amp.getTokenRanges();
+			tokenBscAddress = Amp.getTokenBscAddress();
+			tokenEthereumAddress = Amp.getTokenEthereumAddress();
+		} else {
```
same here?

0xverin committed:
add ERC20TokenLibrary

0xverin committed:
add BRC20TokenLibrary

0xverin committed:
getTokenInfo for erc20

0xverin committed:
getTokenInfo for brc20

wli-pro commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/brc20/BRC20TokenLibrary.sol:
```diff
@@ -0,0 +1,99 @@
+import { Sats } from "./Sats.sol";
+
+library BRC0TokenLibrary {
+	function getBtcsInfo()
```
Would mapping {string => uint256[] ranges} work, i.e. we will  have a mapping (somewhat hard coded, but simplier) like:
{
"sats": Sats.getTokenRanges(),
"rats": Rats.getTokenRanges(),
}

then we can get rid of the hard coded getXyzInfo, instead use the mapping dynamicly map tokenName to get the token's range?

0xverin replied:
like this?:
```contract ERC20Mapping is ERC20{
    constructor(){
        // ada
        tokenNames['ada'] = Ada.getTokenName();
        tokenRanges['ada'] = Ada.getTokenRanges();
        tokenBscAddress['ada'] = Ada.getTokenBscAddress();
        tokenEthereumAddress['ada'] = Ada.getTokenEthereumAddress();
        // amp
        tokenNames['amp'] = Amp.getTokenName();
	tokenNames['amp'] = Amp.getTokenRanges();
	tokenNames['amp'] = Amp.getTokenBscAddress();
	tokenNames['amp'] = Amp.getTokenEthereumAddress();
}

0xverin replied:
Did some changes.

wli-pro commented:
@0xverin Ok I might have potentially found one more gap: the generated VC is lack of the $network info, which is for what networks the VC calculated on, see example https://github.com/litentry/litentry-parachain/pull/2673

0xverin committed:
move if statement

0xverin commented:
> @0xverin Ok I might have potentially found one more gap: the generated VC is lack of the $network info, which is for what networks the VC calculated on, see example #2673

I see in #2785  it has already changed. If fix are needed, shall we address them in the new PR?

0xverin committed:
add mapping for tokens

0xverin committed:
remove unuse setting

0xverin committed:
fix comment

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/libraries/Constants.sol:
```diff
@@ -0,0 +1,23 @@
+pragma solidity ^0.8.8;
+
+library Constants {
+	uint256 constant decimals_factor = 1000;
```
This is only related to token holdings so please move this to `token_holding_amount` directory.

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol:
```diff
@@ -22,33 +22,65 @@ import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.0/contr
+		string memory assertion_type = "Token Holding Amount";
+		schema_url = "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json";
+
+		string memory decodedParams = abi.decode(params, (string));
```
is this token? Could you rename variable ?

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol:
```diff
@@ -22,33 +22,65 @@ import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.0/contr
+			revert("Token not supported or not found");
+		}
+
+		tokenAddresses[Web3Networks.Bsc] = tokenBscAddress[decodedParams];
```
Can this be done in `BRC20.sol` ?

0xverin replied:
Done.

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol:
```diff
@@ -22,33 +22,65 @@ import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.0/contr
-		(uint256 index, uint256 min, int256 max) = calculateRange(balance);
-
-		string[] memory assertions = assembleAssertions(min, max);
+	mapping(uint32 => string) tokenAddresses;
```
Can we have these mappings only for `ERC20` and `BRC20`?

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol:
```diff
@@ -21,62 +21,59 @@ pragma solidity ^0.8.8;
 import "../libraries/Utils.sol";
-import {TokenHoldingAmount} from "./TokenHoldingAmount.sol";
+import { TokenHoldingAmount } from "./TokenHoldingAmount.sol";
+contract BRC20 is TokenHoldingAmount {
```
Can this be abstract ? 

Is this code meant to be uploaded or `BRC20Mapping.sol` ?

0xverin replied:
fixed.

kziemianek reviewed (changes requested):
I though we will have one instance for all token holdings, after reading this PR it turns out that's not the case. It's ok if this is the design you and @higherordertech prefer.

Despite the comments, byte code in repo should be updated so we can see tests pass with bytecode produced from new solidity code.
https://github.com/litentry/litentry-parachain/pull/2811/files#diff-afbca31fd80a6d91ed61556c33e9b8c6e61f6df4b46fa9a91d75b21afb69ac67

0xverin committed:
replace brc20&erc20

0xverin committed:
fix mod bytecode

0xverin committed:
Merge branch 'dev' into refactor-extra-params

0xverin commented:
> I though we will have one instance for all token holdings, after reading this PR it turns out that's not the case. It's ok if this is the design you and @higherordertech prefer.
> 
> Despite the comments, byte code in repo should be updated so we can see tests pass with bytecode produced from new solidity code. https://github.com/litentry/litentry-parachain/pull/2811/files#diff-afbca31fd80a6d91ed61556c33e9b8c6e61f6df4b46fa9a91d75b21afb69ac67

I have not merged brc20 and erc20 temporarily, because:

1. It involves the refactoring of querybalance, and I still don't know the structure of btc solana for now. I want to wait until they are all completed before merging all.

2. Even if we can merge them all, I am worried that it will still affect the use of contracts, as there have been compilation warnings currently:

```
Warning: Contract code size is 38662 bytes and exceeds 24576 bytes (a limit introduced in Spurious Dragon). This contract may not be deployable on mainnet. Consider enabling the optimizer (with a low "runs" value!), turning off revert strings, or using libraries.
  --> contracts/token_holding_amount/erc20/ERC20Mapping.sol:63:1:
  ```

Wdyt? How about we wait for other tokens to be added before merging them together?

0xverin committed:
add erc20 tests

0xverin committed:
fmt

Kailai-Wang commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/libraries/Constants.sol:
duplicate of `tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/Constants.sol`?

Kailai-Wang commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol:
```diff
@@ -17,47 +17,59 @@
+
+		// Initialize network token addresses using Ada library
+
+		
```
Empty comment, or placeholders?

Kailai-Wang reviewed (commented):
Questions:
1. imagine the client adds a new token, does it mean that the `ERC20Mapping.sol` needs to be extended and reployed?
2. as more and more tokens are added, this all-in-one contract will become bigger and bigger. Would it be actually lighter to deploy a separate `tokenX.sol` which inherits the ERC20 contract (so similar to what people do on ethereum).

I'm just tthinking it can be challenging to mantain the all-in-one contract

wli-pro commented:
> Questions:
> 
> 1. imagine the client adds a new token, does it mean that the `ERC20Mapping.sol` needs to be extended and reployed?
> 2. as more and more tokens are added, this all-in-one contract will become bigger and bigger. Would it be actually lighter to deploy a separate `tokenX.sol` which inherits the ERC20 contract (so similar to what people do on ethereum).
> 
> I'm just tthinking it can be challenging to mantain the all-in-one contract

What's the plan in my head is:
1 for initial tokens we migrated we prefer one contract for all, simply because the easier deployment for them.
2 none of our solidity VC is in production usage yet, so this one contract for all way is easier if we find contract defects and we will need fix which need redeploy all of relevant tokens. If the common code is stable, proven defects free, then one contract per token would probably be easier. Before that as it still evolves one contract for all token could be easier to maintain.
3 for new token to be added in future, it's not enforced either way, if VC maintainer feels it's easier for them to add in this one contract or if they feel they want to modify this all token contract to one token contract (that should be easy to do too, or we can provide a placeholder version for that) they can do either of it.

0xverin committed:
merge brc20&&erc20

0xverin committed:
add mod test

0xverin commented:
Updates:

1. Merged ERC20 and BRC20, we only need to deploy the TokenMapping contract

2. Updated dynamic mod test cases, also updated the bytecode of repository

cc @litentry/parachain 
Test evidence:
```
./bin/litentry-cli trusted -d link-identity did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e did:litentry:evm:0x7001f26ebc3e70609736a82ac3f2c8208ffa75ca bsc
 ```
 ```
  ./bin/litentry-cli trusted -d request-vc did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e -a "dynamic 0000000000000000000000000000000000000001 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626e620000000000000000000000000000000000000000000000000000000000"
  ```
  
  ```
  {
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
    ],
    "id": "0xa6e460efcbb9b4e130f6e761030cfbc20671ca36168b2b79cf53eaed8526d434",
    "type": [
        "VerifiableCredential"
    ],
    "credentialSubject": {
        "id": "did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e",
        "description": "The amount of a particular token you are holding",
        "type": "Token Holding Amount",
        "assertionText": "Dynamic(0x0000000000000000000000000000000000000001, BoundedVec([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 98, 110, 98, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 1024))",
        "assertions": [
            {
                "and": [
                    {
                        "src": "$token",
                        "op": "==",
                        "dst": "bnb"
                    },
                    {
                        "src": "$holding_amount",
                        "op": ">=",
                        "dst": "800"
                    },
                    {
                        "src": "$holding_amount",
                        "op": "<",
                        "dst": "1200"
                    }
                ]
            }
        ],
        "values": [
            true
        ],
        "endpoint": "http://localhost:9933/"
    },
    "issuer": {
        "id": "did:litentry:substrate:0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15",
        "name": "Litentry TEE Worker",
        "mrenclave": "2GTxBHhQQyGwYXHRkRAkpchZAmvEJ1JGUw5Xr15Wc11F",
        "runtimeVersion": {
            "parachain": 9181,
            "sidechain": 107
        }
    },
    "issuanceDate": "2024-06-21T05:42:33.614549656+00:00",
    "parachainBlockNumber": 344,
    "sidechainBlockNumber": 681,
    "proof": {
        "created": "2024-06-21T05:42:33.615823708+00:00",
        "type": "Ed25519Signature2020",
        "proofPurpose": "assertionMethod",
        "proofValue": "e9eea72402a3de32a356f4984339b2b5653fa499936a5cffd1f093cf9b8fda5f1fd64fb9a21ffb084a546f377628da65d60f256824acbc980c104b296ade0e0c",
        "verificationMethod": "0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15"
    },
    "credentialSchema": {
        "id": "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json",
        "type": "JsonSchemaValidator2018"
    }
}
```

<img width="1541" alt="image" src="https://github.com/litentry/litentry-parachain/assets/104152026/6408e269-e262-46a0-b5d1-a8157713890d">

0xverin committed:
fix ci

0xverin committed:
fix ci one more

0xverin committed:
Merge branch 'dev' into refactor-extra-params

BillyWooo reviewed (approved):
LGTM
Let's roll it out at least for dev env and test it.

wli-pro reviewed (approved):
Approved this on condition that there is no issue to deploy it to parachain (accept that it might have issues in remix)

wli-pro committed:
Merge branch 'dev' into refactor-extra-params

wli-pro commented:
This PR doesn't make rust interface change but only adjusted / refactored the solidity VC implementation, hence it's not a breaking change to consumer application, hence remove breaking change tag unless any other objection raised.
